qiskit/circuit/library/standard_gates/u.py

```diff
@@ -128,6 +128,35 @@ def __array__(self, dtype=complex):
         )
 
 
+class _CUGateParams(list):
+    # This awful class is to let `CUGate.params` have its keys settable (as
+    # `QuantumCircuit.assign_parameters` requires), while accounting for the problem that `CUGate`
+    # was defined to have a different number of parameters to its `base_gate`, which breaks
+    # `ControlledGate`'s assumptions, and would make most parametric `CUGate`s invalid.
+    #
+    # It's constructed only as part of the `CUGate.params` getter, and given that the general
+    # circuit model assumes that that's a directly mutable list that _must_ be kept in sync with the
+    # gate's requirements, we don't need this to support arbitrary mutation, just enough for
+    # `QuantumCircuit.assign_parameters` to work.
+
+    __slots__ = ("_gate",)
+
+    def __init__(self, gate):
+        super().__init__(gate._params)
+        self._gate = gate
+
+    def __setitem__(self, key, value):
+        if isinstance(key, slice):
+            raise TypeError("'CUGate' only supports setting parameters by single index")
```

Comment:
Does this need a release note? I don't think this was prohibited before, not that I can imagine a use case for it.

Reply:
I can if you like, or I could be less lazy and just fix the logic so it works correctly.  In practice, it's near impossible for a user to do this without breaking the data model of `QuantumCircuit`, so I just did the minimum to support the only safe direct mutation in `QuantumCircuit.assign_parameters` with a noisy error in case I was wrong about all the uses - it didn't trip anything in the library, and `opflow` does some _weird_ stuff.

Reply:
Well I was thinking of something like `cu_gate.params[:3] = [pi, pi/2, -pi]` or something. But we're really in the weeds trying to guess on the interface usage patterns here. I was more just generally thinking about the potential API impact from this change and how to weigh it vs just breaking the API so it's a normal `ControlledGate`.

Reply:
I can fix the logic for the slice, it's not impossible. I was just being lazy, mostly.

Reply:
Done in 4b74b1d57b.